### ISSUE TITLE
Other clustering approaches in the spectral embedding space?

### ISSUE THREAD

**Issue by hbredin** [wq2012/SpectralCluster#30]
As defined in canonical spectral clustering papers, `spectralcluster` uses K-means to perform the final clustering step in the spectral embedding space. 

Yet, K-means is known for not being very good at handling clusters with uneven sizes (which is often the case for speaker diarization, for instance). 

What are your thoughts on this? 
Would it make sense to allow custom clustering approaches in place of K-means?
Or do you think the spectral embedding space is already good enough that it would not make any difference? 

Feel free to close this issue if this is not the place for discussing this...

**Comment by wq2012** [comment]
> Yet, K-means is known for not being very good at handling clusters with uneven sizes (which is often the case for speaker diarization, for instance).

I definitely agree with that. However, I'm also not sure what are the good alternatives we should try. Happy to discuss it here.

If you'd like to offer to expand this library with some alternative algorithms, I'd be happy to merge it and test it with our datasets.

**Comment by hbredin** [comment]
I had no particular alternative in mind but we could try simple stuff like hierarchical agglomerative clustering, GMMs, or any clustering algorithm that can take `num_clusters` as input.

That being said, 
* GMMs might have the same limitations as K-means (regarding uneven clusters)
* HAC (with fixed number of clusters) might be too sensitive to outliers.

Maybe simply expose an additional callable that takes care of the clustering step (and defaults to K-means) ?

**Comment by wq2012** [comment]
Right, I don't think GMM or HAC will work. From our experience they were much worse than K-means.

> Maybe simply expose an additional callable that takes care of the clustering step (and defaults to K-means) ?

Yeah I can do that. No harm to make it more configurable.
